### DISCUSSION TITLE
KP responses given simple reasoning

### DISCUSSION THREAD

**Discussion by cbizon** [NCATSTranslator/TranslatorArchitecture#34] [General]
Our recent KP-related amendment added this language to the architecture doc:

KPs that implement the Translator Reasoner API must perform the following kinds of reasoning in answering queries:
* Making identifiers more specific, e.g. responding to a query involving an entity with information related to a subclass of that entity.
* Making categories in a query more specific. e.g. responding to a query for a biolink:NamedThing with a particular biolink:ChemicalSubstance.
* Making predicates more specific, e.g. responding to a query for “affects expression of” with an edge with predicate “increases expression of”.
* Inverting predicates. e.g. responding to a query with predicate P with an edge whose predicate is the inverse of P.

What does this mean in practice?  

These [slides](https://docs.google.com/presentation/d/1ChrEK7lV0ON8hSeMkNM51j50OyRX47Wi1tv3r-l_Uis/edit#slide=id.p) are an initial guess about answering this question.

As noted on the architecture call, the 'recommendations' there are not based on a coherent principle as much as an idiosyncratic idea of how ARAs would like to use the results.

Thoughts about other approaches?

**Comment by ehinderer** [comment]
For me, intuitively, there's a distinction between cases where the KP contains an _inverted_ predicate and cases where a KP contains an _inferred related_ predicate. In the first case the predicate is essentially equivalent, just with reversed directionality. In the second case, there is a different level semantic interpretation involved. 

In a purely practical sense I don't think an ARA would "care" about a KP returning an inversion for the query (assuming the predicate is truly reflexive) whereas inferred predicate matches could provide valuable information that an ARA could use. 

That being said, and at the risk of the specification being too verbose, I think the simplest and clearest representation of these reasoning steps would be to include an edge attribute in the knowledge graph called "inferred_from" or something similar which points to the original edge in the query graph that the inferred edge was derived.  This could be defaulted to False for those edges that are exact matches. 

The more important question though might be, how do KPs make these inferences? The Connections Hypothesis Provider, for example, does not use ontologies in any capacity at the moment. So it would be pretty difficult for us to enable this kind of functionality. If there's not a shared tool to assist with ontological reasoning, then this requirement would also demand that KPs ingest and reason across ontologies, if I'm understanding this correctly.

**Reply by suihuang-ISB** [reply]
Good point in your last paragraph.  We also struggle with t his distinction that is rarely made: ontological vs. mechanistic (functional-causal, or molecular) relationships or inference thereof.   The latter is what matters when it comes to drug discovery, or explaining diseases etc and I think is more aligned to the original Translator idea. 

The former, ontological and semantic relationships, as you said, indeed is not so much of practical significance, but simply important for sake of interoperability, and using a uniform language - that we adhere to.  Then, mapping from a set of alternative terms that are synonyms  (eg passive/active verb and associated inversion of arrow)  to that which is the commonly agreed upon to use (biolink) is not a matter of "inference" but more elementary: adherence to rules, and in t his sense  the KP would have to do it.

**Reply by cbizon** [reply]
@ehinderer I agree that the how KPs perform this and what kinds of tools do we need to help them is important.  However, I think it's a separable concern, and we should discuss it in a different thread, provided here: https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/32

**Comment by suihuang-ISB** [comment]
Chris: Can you elaborate on the two first bullet points of your examples" for reasoning/inference? They seem quite similar to me (specifying a subclass vs a more specific identifier)...
Also are the three bullet points you list just EXAMPLES to give a sense of the type of reasoning or an EXHAUSTIVE list of THE cases that collectively define the problem of what we mean here by reasoning?

**Reply by cbizon** [reply]
Hi @suihuang-ISB  - I realized I had a typo in the original post, there should have been 4 bullets, not 3.

The difference between the first and second bullet is not huge, but I think it probably matters for implementation.  For bullet 2, all of the inference just involves the biolink model, while bullet 1 involves reasoning over other ontologies such as MONDO.

Maybe the slides linked above make it clearer?  (Unfortunately they're not in the same order as the bullets above, but bullet 1 is on slide 6, while bullet 2 is on slide 4.

These are the exhaustive list of the inferences that every TRAPI KP is expected to provide.  Individual KPs might do more, but this is the baseline requirement.

**Comment by cbizon** [comment]
@ehinderer and others, maybe the basic idea then should be that the KP should provide information as it was asked for, and then include extra nodes/edges in the kgraph that represent the more specific forms from which the asked for edges were derived?  That general form could then be applied across these four cases...

**Reply by ehinderer** [reply]
Well, my point was that maybe the KP _only_ has the more specific predicate, but it is relevant to the query as a subclass of the query predicate. In that case it should just return what information it has and describe in some way that the predicate it returned was different from the one supplied. 

Come to think of it, that description might be better suited for the "edge binding" part of the TRAPI message. But I want to hear others' thoughts on this as well.
